parser_library/src/processing/instruction_sets/macro_processor.cpp

```diff
@@ -182,112 +180,172 @@ context::macro_data_ptr macro_processor::string_to_macrodata(std::string data)
-                    nullptr });
-            }
-            else
-            {
-                if (std::find(keyword_params.begin(), keyword_params.end(), id) != keyword_params.end())
-                {
-                    add_diagnostic(
-                        diagnostic_op::error_E011("Keyword", tmp->operand_range)); // error - keyword already defined
-                }
-                else
-                {
-                    keyword_params.push_back(id);
-                }
-
-                tmp_chain.erase(tmp_chain.begin());
-                tmp_chain.erase(tmp_chain.begin());
-
-                if (tmp_chain.size() == 1 && tmp_chain.front()->type == semantics::concat_type::SUB)
-                    args.symbolic_params.push_back({ mngr.create_macro_data(std::move(tmp_chain), eval_ctx), id });
-                else
-                    args.symbolic_params.push_back(
-                        { string_to_macrodata(mngr.concatenate_str(tmp_chain, eval_ctx)), id });
-            }
-        }
-        else if (tmp->chain.size() == 1 && tmp->chain.front()->type == semantics::concat_type::VAR)
-            args.symbolic_params.push_back({ string_to_macrodata(mngr.convert_to<context::C_t>(
-                                                 mngr.get_var_sym_value(*tmp->chain.front()->access_var(), eval_ctx),
-                                                 tmp->chain.front()->access_var()->symbol_range)),
-                nullptr });
+void macro_processor::get_keyword_arg(const resolved_statement& statement,
+    const semantics::concat_chain& chain,
+    std::vector<context::macro_arg>& args,
+    std::vector<context::id_index>& keyword_params,
+    range op_range) const
+{
+    context_manager mngr(hlasm_ctx);
+
+    auto id = mngr.try_get_symbol_name(chain[0]->access_str()->value).second;
+    assert(id != context::id_storage::empty_id);
+
+    auto named = hlasm_ctx.get_macro_definition(statement.opcode_ref().value)->named_params().find(id);
+    if (named == hlasm_ctx.get_macro_definition(statement.opcode_ref().value)->named_params().end()
+        || named->second->param_type == context::macro_param_type::POS_PAR_TYPE)
+    {
+        add_diagnostic(diagnostic_op::error_E010("keyword parameter", op_range));
+
+        // MACROCASE TODO
+        auto name = chain[0]->access_str()->value;
+
+        args.push_back({ std::make_unique<context::macro_param_data_single>(
+                             name + "=" + semantics::concatenation_point::to_string(chain.begin() + 2, chain.end())),
+            nullptr });
+    }
+    else
+    {
+        if (std::find(keyword_params.begin(), keyword_params.end(), id) != keyword_params.end())
+            add_diagnostic(diagnostic_op::error_E011("Keyword", op_range));
+        else
+            keyword_params.push_back(id);
+
+        auto chain_begin = chain.begin() + 2;
+        auto chain_end = chain.end();
+        auto chain_size = chain.size() - 2;
```

Comment:
what guarantees the size being >=2?

Reply:
it is called only if `is_keyword` predicate is true. The name of this method suggests that the processed operand is a keyword operand and the chain must begin with the `&VAR=`